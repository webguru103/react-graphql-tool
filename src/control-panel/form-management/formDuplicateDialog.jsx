import * as React from 'react';
import type { RouterHistory } from 'react-router';
import { withRouter } from 'react-router';
import TextField from '@material-ui/core/TextField';
import { withStyles } from '@material-ui/core';
import Select from '@material-ui/core/Select';
import MenuItem from '@material-ui/core/MenuItem';
import OutlinedInput from '@material-ui/core/OutlinedInput';
import RadioGroup from '@material-ui/core/RadioGroup';
import FormControl from '@material-ui/core/FormControl';
import FormHelperText from '@material-ui/core/FormHelperText';
import Checkbox from '@material-ui/core/Checkbox';
import Radio from '@material-ui/core/Radio';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import DialogButtonBottom from '../../shared/dialogButtonBottom/dialogButtonBottom';
import styles from './formDuplicateDialog.styles';
import { withFormGroups, withDuplicateForm } from './api/form.service';
import { compose, pro, get } from '../../utility';
import { errParser } from '../../api-error-parser';
import { withDialog } from '../../shared/dialog/withDialog';
import type { Form, FormGroup } from '../../flowTypes';
import type { ParsedApiError } from '../../api-error-parser/errorParser';
import UploadForm from '../../shared/uploadForm/uploadForm';
import c, {
  PDF, MAX_UPLOAD_SIZE,
} from './constants';
import { DEFAULT_PAGE_WIDTH, DEFAULT_PAGE_HEIGHT } from '../../editor/constants';

import type { DocumentUpload, PageUpload } from './flowTypes';

type PropType = {
  duplicateForm: ({
    id: number,
    form: {
      formName: string,
      formGroupId: number,
      formStatus: string,
    },
    sourceURL: string,
    pages: Array<PageUpload>,
  }) => Promise<Form>,
  closeDialog: () => void,
  form: Form,
  classes: Object,
  formGroups: Array<FormGroup>,
  history: RouterHistory,
}

type StateType = {
  networkError: ParsedApiError,
  updateFormBase: boolean,
}

export class FormDuplicateDialogC extends React.PureComponent<PropType, StateType> {

  state = {
    networkError: {},
    updateFormBase: false,
  };

  handleSubmit = async ({
    formName, formGroupId, formStatus, documents,
  }: { formName: string, formGroupId: number, formStatus: string, documents: Array<DocumentUpload> } = {}) => {
    const {
      form: { id }, duplicateForm, closeDialog, history,
    } = this.props;
    const [err, data] = await pro(duplicateForm({
      id,
      form: {
        formName: formName.trim(),
        formGroupId,
        formStatus,
      },
      sourceURL: get(documents[0], 'pdf', null),
      pages: get(documents[0], 'pages', [])
        .map((p) => {
          const newP = {
            ...p,
            sourceURL: p.url || '',
            // TODO: send width and height from the response, right now mocking out sample height/width.
            width: DEFAULT_PAGE_WIDTH,
            height: DEFAULT_PAGE_HEIGHT,
          };
          delete newP.url;
          return newP;
        }),
    }));

    if (err) {
      const parsedError = errParser.parse(err);
      this.setState({ networkError: parsedError });
      return;
    }

    history.push(`/cp-user/editor/edit?formId=${get(data, 'data.form.id', '')}`);
    closeDialog();
  };

  handleUpdateFormBaseToggle = () => {
    this.setState(prevState => ({ updateFormBase: !prevState.updateFormBase }));
  };

  resetNetworkError = () => {
    this.setState({ networkError: {} });
  };

  render() {
    const {
      classes, formGroups, form: {
        formName, formStatus, formGroupId,
      },
    } = this.props;
    const { networkError, updateFormBase } = this.state;
    return (
      <DialogButtonBottom
        actionButtonText={updateFormBase ? 'Upload & Start Editing' : 'Start Editing'}
        onSubmit={this.handleSubmit}
        initialValues={{
          formName: `Copy - ${formName}`,
          formStatus,
          formGroupId,
          updateFormBase: false,
          documents: [],
        }}
        title="Duplicate"
        validate={(values) => {
          const errors = {};
          if (!values.formName) {
            errors.formName = c.NOT_EMPTY;
          }

          if (updateFormBase && values.documents.length === 0) {
            errors.documents = c.NO_FORM_BASE_UPLOADED;
          }
          return errors;
        }}
      >
        {
          ({
            handleChange, values, errors, setErrors, setFieldValue,
          }) => (
            <div className={classes.dialog}>
              <FormControl fullWidth className={classes.formNameInput}>
                <TextField
                  id="form-name"
                  label="Form Name"
                  name="formName"
                  onChange={(e) => { this.resetNetworkError(); handleChange(e); }}
                  margin="normal"
                  error={Boolean(errors.formName || get(networkError, 'validations.formName'))}
                  value={values.formName}
                  fullWidth
                />
                <FormHelperText id="component-error-text" error>{errors.formName || get(networkError, 'validations.formName')}</FormHelperText>
              </FormControl>
              <Select
                value={values.formGroupId}
                onChange={(e) => { this.resetNetworkError(); handleChange(e); }}
                name="formGroupId"
                className={classes.formGroupSelect}
                input={(
                  <OutlinedInput
                    labelWidth={50}
                    fullWidth
                    name="formGroupId"
                    id="formGroupId"
                  />
                )}
              >
                {formGroups.map(fg => <MenuItem key={fg.id} value={fg.id}>{fg.formGroupName}</MenuItem>)}
              </Select>
              <FormControlLabel
                control={(
                  <Checkbox
                    checked={updateFormBase}
                    onChange={this.handleUpdateFormBaseToggle}
                    value="updateFormBase"
                    name="updateFormBase"
                  />
                )}
                label="Update form base"
              />
              {updateFormBase
                && (
                  <FormControl fullWidth>
                    <UploadForm
                      acceptTypes={[PDF]}
                      acceptMultiple={false}
                      maxUploadSize={MAX_UPLOAD_SIZE}
                      uploadLabel={c.UPLOAD_LABEL}
                      setURLs={(response) => { this.resetNetworkError(); setFieldValue('documents', response); }}
                      setUploadError={err => setErrors('documents', err)}
                      onDrop={() => {
                        if (setErrors) {
                          setErrors({
                            ...errors,
                            urls: '',
                          });
                        }
                      }}
                    />
                    <FormHelperText id="upload-error" error>{errors.documents}</FormHelperText>
                  </FormControl>
                )
              }
              <RadioGroup
                aria-label="formStatus"
                name="formStatus"
                className={classes.group}
                value={values.formStatus}
                onChange={(e) => { this.resetNetworkError(); handleChange(e); }}
              >
                <FormControlLabel value="ACTIVE" control={<Radio />} label="Active" />
                <FormControlLabel value="INACTIVE" control={<Radio />} label="Inactive" />
              </RadioGroup>
              {networkError && networkError.global && <FormHelperText error>{networkError.global}</FormHelperText>}
              {networkError && networkError.network && <FormHelperText error>{c.NETWORK_ERROR}</FormHelperText>}
            </div>
          )
        }
      </DialogButtonBottom>
    );
  }

}

export default compose(
  withDuplicateForm,
  withFormGroups,
  withStyles(styles),
  withDialog,
  withRouter,
)(FormDuplicateDialogC);
