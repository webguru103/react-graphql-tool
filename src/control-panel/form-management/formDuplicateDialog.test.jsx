import React from 'react';
import {
  render, fireEvent, cleanup, waitForElement,
} from 'react-testing-library';
import { withContext } from '../../../tests/test-helpers/contextWrapper';
import { FormDuplicateDialogC } from './formDuplicateDialog';

describe('Form Settings Dialog', () => {
  afterEach(cleanup);
  const Component = withContext(FormDuplicateDialogC);
  const mockForm = {
    id: 1,
    formName: 'mockForm',
    formStatus: 'ACTIVE',
    formGroupId: 3,
  };
  const mockFormGroup = {
    id: 1,
    formGroupName: 'mockName',
  };
  const duplicateFormFail = jest.fn(() => Promise.reject({ networkError: 'some' })); //eslint-disable-line
  const updateFormFailGlobal = jest.fn(() => Promise.reject({graphQLErrors: [{ message: 'some', extensions: { model: 'some' } }]})); //eslint-disable-line
  test('renders', () => {
    const { container } = render(<Component classes={{}} form={mockForm} formGroups={[mockFormGroup]} />);
    expect(container).toMatchSnapshot();
  });

  test('Can submit if click on update', (done) => {
    const duplicateForm = jest.fn(() => Promise.resolve());
    const { getByText } = render(
      <Component
        classes={{}}
        form={mockForm}
        formGroups={[mockFormGroup]}
        duplicateForm={duplicateForm}
        history={{
          push: () => {},
        }}
        closeDialog={() => {}}
      />,
    );
    fireEvent.click(getByText('Start Editing'));
    setTimeout(() => {
      expect(duplicateForm).toHaveBeenCalledTimes(1);
      done();
    }, 0);
  });

  test('Should set network error if error is returned by BE', async () => {
    const { getByText } = render(
      <Component
        classes={{}}
        form={mockForm}
        formGroups={[mockFormGroup]}
        duplicateForm={duplicateFormFail}
        closeDialog={() => {}}
      />,
    );
    await fireEvent.click(getByText('Start Editing'));
    const errorText = await waitForElement(() => getByText('Some network error occurred, try again later.'));
    expect(errorText).toBeDefined();
  });

  test('Cannot submit if the formName is empty', () => {
    const duplicateForm = jest.fn(() => Promise.resolve());
    const { getByText, getByLabelText } = render(
      <Component
        classes={{}}
        form={mockForm}
        formGroups={[mockFormGroup]}
        duplicateForm={duplicateForm}
        closeDialog={() => {}}
        createSnackbar={() => {}}
      />,
    );
    fireEvent.change(getByLabelText('Form Name'), {
      target: { value: '' },
    });
    fireEvent.click(getByText('Start Editing'));
    expect(duplicateForm).toHaveBeenCalledTimes(0);
  });

  test('Reset the network error on radio group switch', async (done) => {
    const { getByText, getByLabelText, queryByText } = render(
      <Component
        classes={{}}
        form={mockForm}
        formGroups={[mockFormGroup]}
        duplicateForm={duplicateFormFail}
        closeDialog={() => {}}
      />,
    );
    await fireEvent.click(getByText('Start Editing'));
    await waitForElement(() => getByText('Some network error occurred, try again later.'));
    expect(getByText('Some network error occurred, try again later.')).toBeDefined();
    await fireEvent.click(getByLabelText('Inactive'));
    setTimeout(() => {
      expect(queryByText('Some network error occurred, try again later.')).toBeNull();
      done();
    }, 0);
  });
});
