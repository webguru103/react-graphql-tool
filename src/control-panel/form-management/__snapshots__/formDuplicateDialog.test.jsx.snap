// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Settings Dialog renders 1`] = `
<div>
  <form>
    <div
      class="MuiDialogTitle-root-7 DialogCreatorInjector-dialogTitle-2"
    >
      <span>
        Duplicate
      </span>
      <button
        class="MuiButtonBase-root-14 MuiIconButton-root-8 DialogCreatorInjector-dialogClose-6"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label-13"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root-17"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
            />
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
          </svg>
        </span>
      </button>
    </div>
    <div
      class="MuiDialogContent-root-26 DialogCreatorInjector-dialogContent-3"
    >
      <div>
        <div
          class="MuiFormControl-root-27 MuiFormControl-fullWidth-30"
        >
          <div
            class="MuiFormControl-root-27 MuiFormControl-marginNormal-28 MuiFormControl-fullWidth-30"
          >
            <label
              class="MuiFormLabel-root-38 MuiFormLabel-filled-42 MuiInputLabel-root-31 MuiInputLabel-formControl-32 MuiInputLabel-animated-35 MuiInputLabel-shrink-34"
              data-shrink="true"
              for="form-name"
            >
              Form Name
            </label>
            <div
              class="MuiInputBase-root-58 MuiInput-root-45 MuiInput-underline-49 MuiInputBase-fullWidth-67 MuiInput-fullWidth-52 MuiInputBase-formControl-59 MuiInput-formControl-46"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input-68 MuiInput-input-53"
                id="form-name"
                name="formName"
                type="text"
                value="Copy - mockForm"
              />
            </div>
          </div>
          <p
            class="MuiFormHelperText-root-75 MuiFormHelperText-error-76"
            id="component-error-text"
          />
        </div>
        <div
          class="MuiInputBase-root-58 MuiOutlinedInput-root-90 MuiInputBase-fullWidth-67"
        >
          <fieldset
            aria-hidden="true"
            class="MuiNotchedOutline-root-103 MuiOutlinedInput-notchedOutline-97"
            style="padding-left: 30.75px;"
          >
            <legend
              class="MuiNotchedOutline-legend-104"
              style="width: 0.01px;"
            />
          </fieldset>
          <div
            class="MuiSelect-root-83"
          >
            <div
              aria-haspopup="true"
              aria-pressed="false"
              class="MuiSelect-select-84 MuiSelect-selectMenu-87 MuiInputBase-input-68 MuiOutlinedInput-input-98"
              role="button"
              tabindex="0"
            >
              <span>
                â€‹
              </span>
            </div>
            <input
              id="formGroupId"
              name="formGroupId"
              type="hidden"
              value="3"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root-17 MuiSelect-icon-89"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
        </div>
        <label
          class="MuiFormControlLabel-root-112"
        >
          <span
            class="MuiButtonBase-root-14 MuiIconButton-root-8 MuiSwitchBase-root-122 MuiCheckbox-root-116 MuiCheckbox-colorSecondary-121"
          >
            <span
              class="MuiIconButton-label-13"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root-17"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
              </svg>
              <input
                class="MuiSwitchBase-input-125"
                data-indeterminate="false"
                name="updateFormBase"
                type="checkbox"
                value="updateFormBase"
              />
            </span>
            <span
              class="MuiTouchRipple-root-126"
            />
          </span>
          <span
            class="MuiTypography-root-133 MuiTypography-body1-142 MuiFormControlLabel-label-115"
          >
            Update form base
          </span>
        </label>
        <div
          aria-label="formStatus"
          class="MuiFormGroup-root-159"
          role="radiogroup"
        >
          <label
            class="MuiFormControlLabel-root-112"
          >
            <span
              class="MuiButtonBase-root-14 MuiIconButton-root-8 MuiSwitchBase-root-122 MuiRadio-root-161 MuiRadio-colorSecondary-165 MuiSwitchBase-checked-123 MuiRadio-checked-162"
            >
              <span
                class="MuiIconButton-label-13"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root-17"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 7c-2.76 0-5 2.24-5 5s2.24 5 5 5 5-2.24 5-5-2.24-5-5-5zm0-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                  />
                </svg>
                <input
                  checked=""
                  class="MuiSwitchBase-input-125"
                  name="formStatus"
                  type="radio"
                  value="ACTIVE"
                />
              </span>
              <span
                class="MuiTouchRipple-root-126"
              />
            </span>
            <span
              class="MuiTypography-root-133 MuiTypography-body1-142 MuiFormControlLabel-label-115"
            >
              Active
            </span>
          </label>
          <label
            class="MuiFormControlLabel-root-112"
          >
            <span
              class="MuiButtonBase-root-14 MuiIconButton-root-8 MuiSwitchBase-root-122 MuiRadio-root-161 MuiRadio-colorSecondary-165"
            >
              <span
                class="MuiIconButton-label-13"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root-17"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                  />
                </svg>
                <input
                  class="MuiSwitchBase-input-125"
                  name="formStatus"
                  type="radio"
                  value="INACTIVE"
                />
              </span>
              <span
                class="MuiTouchRipple-root-126"
              />
            </span>
            <span
              class="MuiTypography-root-133 MuiTypography-body1-142 MuiFormControlLabel-label-115"
            >
              Inactive
            </span>
          </label>
        </div>
      </div>
    </div>
    <div
      class="MuiDialogActions-root-166 DialogCreatorInjector-dialogActions-4"
    >
      <button
        class="MuiButtonBase-root-14 MuiButton-root-172 MuiButton-contained-183 MuiButton-containedSecondary-185 ButtonC-secondary-170 MuiButton-raised-186 MuiButton-raisedSecondary-188 ButtonC-button-168"
        data-testid="close-dialog-button-bottom"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label-173"
        >
          Cancel
        </span>
      </button>
      <button
        class="MuiButtonBase-root-14 MuiButton-root-172 MuiButton-contained-183 MuiButton-containedPrimary-184 ButtonC-primary-169 MuiButton-raised-186 MuiButton-raisedPrimary-187 ButtonC-button-168"
        data-testid="action-dialog-button-bottom"
        tabindex="0"
        type="submit"
      >
        <span
          class="MuiButton-label-173"
        >
          Start Editing
        </span>
      </button>
    </div>
  </form>
</div>
`;
